libusb/os/linux_usbfs.c

```diff
@@ -1407,6 +1408,129 @@ static int op_wrap_sys_device(struct libusb_context *ctx,
 	return r;
 }
 
+static int check_subsystem(const char *sys_path, const char* subsystem)
+{
+	char *path, *subsystem_path;
+	int ret;
+
+	ret = asprintf(&path, "%s/subsystem", sys_path);
+	if (ret < 0)
+		return LIBUSB_ERROR_NO_MEM;
+
+	subsystem_path = realpath(path, NULL);
+	free(path);
+	if (!subsystem_path && errno == ENOMEM)
+		return LIBUSB_ERROR_NO_MEM;
+
+	if (subsystem_path) {
+		ret = !!strcmp(subsystem_path, subsystem);
+		free(subsystem_path);
+	}
+
+	return ret;
+}
+
+static int get_subsytem(struct libusb_context *ctx, char *buf, size_t len,
+	const char *dir, const char* subsystem, int depth)
+{
+	DIR *dp;
+	struct dirent *entry;
+	struct stat statbuf;
+	char *path;
+	int ret;
+
+	if (depth >= 20)
+		return LIBUSB_ERROR_NOT_FOUND;
+
+	ret = check_subsystem(dir, subsystem);
+	if (ret < 0)
+		return ret;
+
+	if (!ret) {
+		path = strrchr(dir, '/');
+		if(path) {
+			snprintf(buf, len, "/dev/%s", path + 1);
```

Comment:
Maybe I need more coffee, but if `/` is the very last character in the string, `path + 1` is pointing to the null terminator, right?

Reply:
`strrchr(dir, '/')` will be `/sde` as in it incldes the `/` if its not found it will be `NULL`.

Oh im dumb i can just do
`snprintf(buf, len, "/dev%s", path);` anyway. i was returning `sde` but changed it oops

Reply:
That formulation is indeed better.

But what I was getting at was: what if (somehow) `/` is the *very last* character in `dir`?  Then `buf` will become `/dev/` and you'll return that. But that's a bogus result. Better to fail.  I think you need to also check the `strrchr` result (`path`) not just for NULL but `strlen(path) > 1`.

Reply:
`dir` can not end in a `/` as it comes from `1469: ret = asprintf(&path, "%s/%s", dir, entry->d_name);`. `path` in this case will always be the caller's `/` + `entry->d_name`.

Reply:
That's where it comes from today. But the correctness of the `check_subsystem` function should stand on its own. Imagine some future refactoring from someone that doesn't have the flow fresh in his mind and calls `check_subsystem` from some place new (hence the "(somehow)" in my previous comment).  Or imagine libusb someday having great testing where a fuzzer is generating crazy inputs.